Adding run.py

BenjaminBossan opened the pull request:
From the `method_comparison/causal_lm` directory:

```
python run.py -v experiments/lora/rank32/
```

BenjaminBossan committed:
Empty run.py

BenjaminBossan committed:
First implementation of training script

Still WIP:

- no sacred or other experiment tracking framework
- only toy task

githubnemo commented on method_comparison/causal_lm/temporary_results/lora--rank32--2025-02-25T16-39-53+00-00.json:
```diff
@@ -0,0 +1,47 @@
+      5.82646369934082,
+      3.189246892929077,
+      3.522594690322876
+    ]
```
parameters are not tracked here - is this expected?

BenjaminBossan replied:
You mean the `training_params.json` and `adapter_config.json`? I didn't add them, I only added their hashes.

githubnemo replied:
For rendering it might be beneficial to include them. WDYT?

BenjaminBossan replied:
Done. I removed the hashes, as I don't think there is a point in having both.

BenjaminBossan committed:
Use MetaMathQA, update scripts

BenjaminBossan committed:
Multiple fixes and improvements

- Error with data processing, now learns properly
- Use tqdm
- Improve console logging
- Adjust data filtering criteria
- Add option for cosine lr scheduler

BenjaminBossan committed:
Update docs, .gitignore, requirements

BenjaminBossan committed:
Add option to use automatic mixed precision

Off by default as it gives OOM without any speed benefit.

BenjaminBossan committed:
Move all logging stuff to utils.py

BenjaminBossan committed:
Some fixes suggested by mypy

BenjaminBossan committed:
Some formatting

BenjaminBossan committed:
Improve docs

BenjaminBossan committed:
Include one EOS token in loss calculation

Also, model never predicts EOS.

BenjaminBossan committed:
Fix to progress bar

BenjaminBossan committed:
Log tokens / sec

BenjaminBossan committed:
Add default training params

The experiment specific training params use the default training params
but can override any parameter from it if needed. However, this way it's
easier to make a change to all experiments (say, I want to change the
base model, I don't need to change each individual
training_parameters.json).

BenjaminBossan committed:
Add possibility to change attn implementation

However, both flash attention 2 and flex attention are slower on my
system. Thus, stay with default None (-> SDPA).

BenjaminBossan committed:
Fix bugs when there is no training_params.json

BenjaminBossan committed:
Refactor to use GenerationConfig

Allows to more easily use, say, static cache, which is the new default,
as it's faster (apart from the first pass)

BenjaminBossan committed:
Support for separate eval batch size

BenjaminBossan committed:
Small fixes, slightly lower char limit

BenjaminBossan committed:
Support running in offline mode

BenjaminBossan committed:
Also log reserved memory

BenjaminBossan committed:
Small fixes

Type annotations, a training param, comments

BenjaminBossan committed:
Log whole train and peft config

No longer log the sha256sum of them

BenjaminBossan committed:
Update README to reflect latest changes in results

Also, print number of (trainable) parameters.

BenjaminBossan committed:
Small adjustments to support AdaLoRA

BenjaminBossan committed:
Update READMEs

- Clarify dataset split for the MetaMathQA task
- Add section on contributing

BenjaminBossan committed:
Slight refactoring of data loading functions

BenjaminBossan committed:
Change valid/test set to use GSM8K

BenjaminBossan committed:
Better parsing of answers

E.g. 1/2 == 0.5

BenjaminBossan committed:
Small updates to README

BenjaminBossan committed:
Add a couple of experiment configs

BenjaminBossan committed:
Add more stuff to logging

BenjaminBossan committed:
Improve eval speed by setting dynamic max length

BenjaminBossan committed:
Add small Gradio demo

BenjaminBossan committed:
Update to app requirements

BenjaminBossan committed:
Small update to processing and app, prettier

BenjaminBossan committed:
Add option to filter the df

BenjaminBossan committed:
Introduce BucketIterator for faster training

BenjaminBossan committed:
Update ideas for interesting tasks

BenjaminBossan committed:
Keep adapter file by default after train run

But add --clean to delete it.

Keeping the adapter can be useful if the user wants to run further tests
with the trained model.

BenjaminBossan committed:
Move imports from local to global

Otherwise, if one of these packages are missing, we could get an error
after the model training and lose all progress.

githubnemo commented on method_comparison/MetaMathQA/utils.py:
```diff
@@ -241,6 +241,51 @@ def get_model(
+        num_samples = len(bucket["input_ids"])
+        for j in range(0, num_samples, self.batch_size):
+            batch = {k: v[j : j + self.batch_size] for k, v in bucket.items()}
+            yield type(bucket)(batch)
```
`type(bucket)` will always be dict here, is this intended?

BenjaminBossan replied:
Will it? I wanted to be careful because sometimes we get dict-like objects from HF libraries (like the output from `forward` of a transformers model with `return_dict=True`), so I wanted to ensure the same type remains. If it's still dict, it won't hurt except for a minimal performance overhead, right?

BenjaminBossan committed:
Small fixes to AMP

BenjaminBossan committed:
Small fix for bucket type
